### PR TITLE
file-seach: improve handling of command

### LINKED ISSUE #7844: file-search: better handling of command when no workspace is present
#### Bug Description:

There is an error thrown when attempting to execute `file-search` without any workspace opened.
There should be better handling for such a case:

<div align='center'>

Screen Shot 2020-05-19 at 11 36 16 AM

</div>

#### Steps to Reproduce:

1. open the application without any workspace present (or close the workspace)
2. execute the 'file-search' command <kbd>ctrlcmd</kbd>+<kbd>p</kbd>
3. verify the logs, the following error will be thrown:

```
logger-protocol.ts:112 root ERROR Failed to execute command: Error: The command 'file-search.openFile' cannot be executed. There are no active handlers available for the command. (args: [null])
    at CommandRegistry.<anonymous> (file:///Users/vincentfugnitto/workspace/theia/examples/electron/lib/bundle.js:128506:45)
    at step (file:///Users/vincentfugnitto/workspace/theia/examples/electron/lib/bundle.js:128275:23)
    at Object.next (file:///Users/vincentfugnitto/workspace/theia/examples/electron/lib/bundle.js:128256:53)
    at file:///Users/vincentfugnitto/workspace/theia/examples/electron/lib/bundle.js:128250:71
    at new Promise (<anonymous>)
    at ../../packages/core/lib/common/command.js.__awaiter (file:///Users/vincentfugnitto/workspace/theia/examples/electron/lib/bundle.js:128246:12)
    at CommandRegistry.../../packages/core/lib/common/command.js.CommandRegistry.executeCommand (file:///Users/vincentfugnitto/workspace/theia/examples/electron/lib/bundle.js:128488:16)
    at KeybindingRegistry.../../packages/core/lib/browser/keybinding.js.KeybindingRegistry.executeKeyBinding (file:///Users/vincentfugnitto/workspace/theia/examples/electron/lib/bundle.js:106007:38)
    at KeybindingRegistry.../../packages/core/lib/browser/keybinding.js.KeybindingRegistry.run (file:///Users/vincentfugnitto/workspace/theia/examples/electron/lib/bundle.js:106102:22)
    at HTMLDocument.<anonymous> (file:///Users/vincentfugnitto/workspace/theia/examples/electron/lib/bundle.js:104280:35)
```

### PR DESCRIPTION
**Pull request by vince-fugnitto**
#### What it does

Fixes #7844

This pull-request contributes the following items:
- introduces the `workspaceState` when context which can ultimately be used by new commands
- improves the handling of the **quick-file open**:
   - enables the command in all cases (even when no workspace is opened)
   - adds proper handling when no results are found (`No matching results`) 

#### How to test

1. start the application without a workspace opened
2. execute the `file-search` command (<kbd>ctrlcmd</kbd>+<kbd>p</kbd>)
3. the quick-file-open should be properly displayed
4. searching for results will yield the `No matching results` item
5. open a workspace
6. execute the `file-search` command (<kbd>ctrlcmd</kbd>+<kbd>p</kbd>)
7. searching should yield results in the quick-file open
8. search for a random string (ex: `adsaasdadsadad`) it should yield the `No matching results` item

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: vince-fugnitto <vincent.fugnitto@ericsson.com>

### CODE REVIEW

File: packages/file-search/src/browser/quick-file-open-contribution.ts [hunk 1]
```diff
@@ -47,7 +48,8 @@ export class QuickFileOpenFrontendContribution implements CommandContribution, K
     registerKeybindings(keybindings: KeybindingRegistry): void {
         keybindings.registerKeybinding({
             command: quickFileOpen.id,
-            keybinding: 'ctrlcmd+p'
+            keybinding: 'ctrlcmd+p',
+            context: WorkspaceKeybindingContexts.workspaceOpened
```
Comment (akosyakov) [inline packages/file-search/src/browser/quick-file-open-contribution.ts:52]: we should use `when` instead, `context` is deprecated and will be removed

Comment (akosyakov) [reply packages/file-search/src/browser/quick-file-open-contribution.ts:52]: There is a new when context in VS Code: https://github.com/microsoft/vscode-docs/commit/3ca603d30295db5728d6f403bfd7b0f405264737

We should implement it in https://github.com/eclipse-theia/theia/blob/f4a92a7c8fd15140f542cca55d2d67441158a819/packages/workspace/src/browser/workspace-frontend-contribution.ts#L57-L66

and then use as `when: 'workspaceState != empty'`

Comment (vince-fugnitto) [reply packages/file-search/src/browser/quick-file-open-contribution.ts]: > we should use `when` instead, `context` is deprecated and will be removed

Thank you for the suggestion and guidance, I updated the code to instead implement and use the `workspaceState` when clause.

**Comment by vince-fugnitto** [comment]
@lmcbout 

> Somewhere in the issue, it should stipulate this is for ELECTRON only, not for the browser. If there is no workspace and I do (CTRL + P) in the browser, there is a command attach to it.(Print)

There is no **electron**-specific code present in the pull-request.
The pull-request introduces the `workspaceState` when-clause which can be used irregardless of the target. 

> Works as expected BUT if I open the command and remove the ">", it says in the quick open bar "File name to search" , but it is looking inside the workspace and if there is no workspace, it does nothing. (As an end-user, I expect something here) Should we allow to have the quick open search box available in this case?

This is consistent with vscode and I do not want to differ the behavior in this pull-request.
I don't think this behavior is confusing, logically you will not be able to search without files present.

> 
> Second, If I open a file (CTRL + O) with no workspace, it opens the file and then the command (CTRL+P) it opens the quick search , but still, cannot return any files to open.

It searches for files present in a workspace (a single file is not a workspace).

**Comment by lmcbout** [comment]
So if not exclusive to ELECTRON, then it does not work properly since the command ( CTRL + P) in a CHROME browser, it tries to **print**, not opening the quick open file selection

**Comment by vince-fugnitto** [comment]
> So if not exclusive to ELECTRON, then it does not work properly since the command ( CTRL + P) in a CHROME browser, it tries to **print**, not opening the quick open file selection

@lmcbout the command should not be present without a workspace present (which is the case).
The <kbd>ctrl</kbd>+<kbd>p</kbd> is a reserved keybinding (ex: chrome) if no command is present.

**Comment by vince-fugnitto** [comment]
@akosyakov should we remove the `print` behavior when no command is available which overrides <kbd>ctrlcmd</kbd>+<kbd>p</kbd>?

**Comment by lmcbout** [comment]
When opening the quick open search box, the text inside the search field should be 
From:
File name to search ...
To:
File name to search in current workspace ....

**Comment by vince-fugnitto** [comment]
> When opening the quick open search box, the text inside the search field should be
> From:
> File name to search ...
> To:
> File name to search in current workspace ....

@lmcbout why?

I have a couple of issues with your suggestion:
- modifying the placeholder is out of the scope of the pull-request and is feature-creeping.
- modifying the placeholder to the value you proposed does not align with vscode, they have a different placeholder than us but they don't mention 'in the current workspace'
- I don't see the value for adding 'in the current workspace' as it is implied.

**Comment by vince-fugnitto** [comment]
@akosyakov what should the behavior be when we are in the `browser` and no command is available which uses the browser's print keybinding? At the moment (after these changes), the print dialog from the browser will be prompted to end-users.

**Comment by akosyakov** [comment]
@vince-fugnitto VS Code shows the quick find always, so no fix? introducing a new context key is nice anyway

Screenshot 2020-05-25 at 10 02 11

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto VS Code shows the quick find always, so no fix? introducing a new context key is nice anyway

I can align to show the context-menu. 
The issue of displaying the browser's **print** will still occur for applications which are not using the `file-search` however, so I was wondering if there was anything we should do about it.

**Comment by akosyakov** [comment]
> I can align to show the context-menu.

@vince-fugnitto Not sure what you mean. In VS Code they always show the quick search, we should do the same, just fix the error in console. If we show it always then there is no print.

**Comment by vince-fugnitto** [comment]
> > I can align to show the context-menu.
> 
> @vince-fugnitto Not sure what you mean. In VS Code they always show the quick search, we should do the same, just fix the error in console. If we show it always then there is no print.

Yes I’m currently working on it.
What I meant was that it an application does not include `@theia/file-search` as an extension then using the <kbd>ctrlcmd</kbd>+<kbd>p</kbd> keybinding will result in the browser’s print window. I’m asking if in a separate pull-request we should override this behaviour (ex: in core) so the print dialog is not shown.

**Comment by akosyakov** [comment]
> What I meant was that it an application does not include @theia/file-search as an extension then using the ctrlcmd+p keybinding will result in the browser’s print window.

Let's wait for issues from users about it.

**Comment by vince-fugnitto** [comment]
@lmcbout do you mind retrying the pull-request?

**Comment by lmcbout** [comment]
@vince-fugnitto will do in a few minutes.

**Comment by vince-fugnitto** [comment]
> Note: may be on this PR or on another PR: If we use "CTRLCMD+P", -> open the quick view, then I enter the whole path for the file, we should allow to open that file even if there is no workspace.
> i.e. "CTRLCMD+P" then enter "/home/lmcbout/.bashrc" (I am on UBUNTU, so this file exist for me), the file should open in the editor

@lmcbout this is not the purpose of the quick-file open (it does not search for files outside the workspace nor does it open files on the filesystem by path). This behavior also does not align with vscode so I am not for adding it.

**Comment by lmcbout** [comment]
>@lmcbout this is not the purpose of the quick-file open (it does not search for files outside the workspace nor does it open files on the filesystem by path). This behavior also does not align with vscode so I am not for adding it.

Just did a quick test on VSCODE and the file opens, so I guess it would be aligned with VSCODE :+1:

**Comment by vince-fugnitto** [comment]
> > @lmcbout this is not the purpose of the quick-file open (it does not search for files outside the workspace nor does it open files on the filesystem by path). This behavior also does not align with vscode so I am not for adding it.
> 
> Just did a quick test on VSCODE and the file opens, so I guess it would be aligned with VSCODE 👍

Not in this pull-request.

**Pull request merged**